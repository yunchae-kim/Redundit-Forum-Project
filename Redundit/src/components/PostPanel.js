/* eslint-disable max-len */
import { Divider } from "@blueprintjs/core";
import { useEffect, useState } from "react";
import { Link, useHistory } from "react-router-dom";
import userIcon from "../resources/userIcon.png";
import { getPostById, deletePostById, flagPostById } from "../utility/axiosPost";
import { addHide } from "../utility/axiosHide";
import { getMember } from "../utility/axiosMember";
import { getHashtag } from "../utility/axiosHashtags";
import "../style/components/PostPanel.css";
import { addNotification } from "../utility/axiosNotification";

const PostPanel = ({ postId, userId }) => {
  const [post, setPost] = useState([]);
  const [fetched, setfetched] = useState(false);
  const [member, setMember] = useState(null);
  const [tags, setTags] = useState([]);
  const history = useHistory();

  useEffect(() => {
    setfetched(true);
    getPostById(postId).then((postFetched) => {
      setPost(postFetched);
      getMember(postFetched.idgroup, userId).then((memberFetched) => {
        setMember(memberFetched);
      });
    });
  }, [fetched]);

  useEffect(() => {
    getHashtag(postId, "post").then((tagsFetched) => {
      setTags(tagsFetched);
    });
  }, []);

  const hashTagsList = tags.map((h) => (
    <Link to={{ pathname: `/hashtag/${h.hashtag}`, state: { hashtag: h.hashtag } }} key={h.hashtag}>
      #
      {h.hashtag}
      {" "}
    </Link>
  ));

  const deletePost = () => {
    deletePostById(postId).then(() => {
      if (member !== null && member.admin === 1) {
        addNotification(post.iduser, `Your post ${post.title} is deleted by admin.`).then(() => {
          history.push({ pathname: `/group/${post.idgroup}`, state: { groupId: post.idgroup } });
          history.go();
        });
      } else {
        history.push({ pathname: `/group/${post.idgroup}`, state: { groupId: post.idgroup } });
        history.go();
      }
    });
  };

  const editPost = () => {
    history.push({ pathname: `/editpost/post/${postId}`, state: { post, hashtags: tags } });
  };

  const hidePost = () => {
    addHide(userId, postId, post.idgroup).then(() => {
      history.push({ pathname: `/group/${post.idgroup}`, state: { groupId: post.idgroup } });
      history.go();
    });
  };

  const flagPost = () => {
    flagPostById(postId, true);
  };

  return (
    <div className="PostPanel">
      <div className="PP-header">
        <img className="PP-userIcon" src={userIcon} alt="" />
        <div className="PP-userName">
          {post.status === 1
            ? <Link to={{ pathname: `/user/${post.iduser}`, state: { userId: post.iduser } }}><span className="CW-chatentry-name">{post.username}</span></Link>
            : (
              <span className="CW-chatentry-name">
                {post.username}
                {" "}
                (deactivated)
              </span>
            )}
        </div>
        <div className="PP-postDate">
          Posted:
          {" "}
          {post.date}
        </div>
      </div>
      <div className="PP-title">
        {post.title}
      </div>
      <div className="PP-hashtags">
        {hashTagsList}
      </div>
      <div className="PP-post">
        {post.content}
      </div>
      <div className="PP-viewPost">
        <Link to={{ pathname: `/post/${postId}`, state: { postId } }} className="PP-viewPostText">
          View More
        </Link>
      </div>
      <hr />
      <div className="PP-fooder">
        {((userId !== 0 && userId !== post.iduser)) && <button type="submit" onClick={() => hidePost()}>Hide</button>}
        <Divider />
        {((member != null)) && <button type="submit" onClick={() => flagPost()}>Flag for Deletion</button>}
        <Divider />
        {((post.iduser === userId) || (member != null && member.admin === 1)) && <button type="submit" onClick={() => deletePost()}>Delete</button>}
        <Divider />
        {(post.iduser === userId) && <button type="submit" onClick={() => editPost()}>Edit</button>}
      </div>
    </div>
  );
};

export default PostPanel;
