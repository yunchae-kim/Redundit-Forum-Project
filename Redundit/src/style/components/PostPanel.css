.PostPanel {
    width: 1100px;
    height: 400px;
    margin: 29px 100px 0 160px;
    padding: 17px 1px 21px 0;
    border-radius: 20px;
    background-color: #fff;
    font-family: HelveticaNeue;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    line-height: normal;
    letter-spacing: normal;
    text-align: left;
  }
  
  .PP-header {
    display: flex;
    flex-direction: row;
    margin-left: 30px;
    margin-right: 32px;
    padding-top: 0px;
    padding-bottom: 10px;
    color: black;
    font-size: 16px;
  }
  
  .PP-userIcon {
    border-radius: 30%;
    width: 20px;
    height: 20px;
    align-self: center;
    margin-right: 15px;
  }
  
  .PP-postDate {
    align-self: center;
  }
  .PP-userName {
    margin-right: 30px;
    align-self: center;
    font-weight: bold;
  }
  .PP-title {
    height: 28px;
    margin: 0px 167px 10px 30px;
    font-family: HelveticaNeue;
    font-size: 24px;
    font-weight: bold;
    color: #111;
  }
  
  .PP-hashtags {
    margin: 10px 318px 16px 30px;
    font-size: 16px;
    color: #465771;
  }
  
  .PP-post {
    width: 1038px;
    height: 140px;
    margin: 15px 36px 15px 30px;
    font-size: 20px;
    overflow: hidden;
    text-overflow: ellipsis;
    color: #111;
  }
  
  .PP-fooder {
    display: flex;
    width: 485px;
    height: 18px;
    margin: 15.5px 587px 0 30px;
    font-size: 16px;
    color: #188bd8;
  }
  
  .PP-viewPost{
    margin-top: 20px;
    margin-bottom: 20px;
    font-size: 20px;
    font-weight: bold;
    text-align: center;
    align-self: center;
  }
  
  .PP-viewPostText {
    color : #5da271;
  }