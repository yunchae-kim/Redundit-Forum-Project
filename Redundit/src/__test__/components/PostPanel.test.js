import {
  waitFor,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import PostPanel from "../../components/PostPanel";

test("test should PostPanel match snapshot", async () => {
  const tree = renderer.create(
    <BrowserRouter>
      <PostPanel postId={1} userId={1} />
    </BrowserRouter>,
  ).toJSON();
  await waitFor(async () => {
    expect(tree).toMatchSnapshot();
  });
});
