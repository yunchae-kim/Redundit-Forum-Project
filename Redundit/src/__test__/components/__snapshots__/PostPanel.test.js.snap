// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should PostPanel match snapshot 1`] = `
<div
  className="PostPanel"
>
  <div
    className="PP-header"
  >
    <img
      alt=""
      className="PP-userIcon"
      src="userIcon.png"
    />
    <div
      className="PP-userName"
    >
      <span
        className="CW-chatentry-name"
      >
         
        (deactivated)
      </span>
    </div>
    <div
      className="PP-postDate"
    >
      Posted:
       
    </div>
  </div>
  <div
    className="PP-title"
  />
  <div
    className="PP-hashtags"
  />
  <div
    className="PP-post"
  />
  <div
    className="PP-viewPost"
  >
    <a
      className="PP-viewPostText"
      href="/post/1"
      onClick={[Function]}
    >
      View More
    </a>
  </div>
  <hr />
  <div
    className="PP-fooder"
  >
    <button
      onClick={[Function]}
      type="submit"
    >
      Hide
    </button>
    <div
      className="bp3-divider"
    />
    <div
      className="bp3-divider"
    />
    <div
      className="bp3-divider"
    />
  </div>
</div>
`;
